README.md
# FlaskJWE

## Introduction

FlaskJWE provides a very simple way to integrate JSON Web Encryption (JWE) as defined in [RFC 7516](https://tools.ietf.org/html/rfc7516)
into your Flask server. This plugin allows the server to accept JWE requests and optionally (but by default) respond with 
JWE responses. This module depends on the [pyjwkest](https://github.com/rohe/pyjwkest), so the algorithms and encoding methods 
supported by this module are those that are supported by **pyjwkest**.

## Dependecies

1. [pyJwkest](https://github.com/rohe/pyjwkest)
2. [PyRedis](http://pyredis.readthedocs.org/en/latest/)

## How to Use

In order to allow your Flask server to accept or respond to JWE messages, simply instantiate the plugin on the Flask app:

```python
from flask import Flask
from flask.ext.jwe import FlaskJWE

app = Flask('MyTestApp')
encrypted_content = FlaskJWE(app)
```

In order to do some advanced configuration

```python
from flask import Flask
from flask.ext.jwe import FlaskJWE

app = Flask('MyTestApp')
encrypted_content = FlaskJWE(app)

app.config['JWE_REDIS_URI'] = 'redis://localhost:6379/1'                        # Default
app.config['SERVER_PUB_JWK_ENDPOINT'] = '/serverpubkey'                         # Default
app.config['JWE_ECDH_ES'] = True                                                # Default
app.config['ECDH_CURVE'] = 'P-256'                                              # Default, other options: 'P-256', 'P-384' or 'P-512'
app.config['JWE_ES_KEY_EXPIRES'] = 600                                          # Not Default
app.config['JWE_SERVER_KEY_PER_IP'] = True                                      # Default
app.config['JWE_KEY_ENCRYPTION_KEY'] = int(os.urandom(32).encode('hex'), 16)    # Not Default
app.config['JWE_SET_REQUEST_DATA'] = True                                       # Default

symkey = SYMKey(key='supersecretsymmetrickey')
app.config['JWE_SERVER_SYM_KEY'] = symkey                   # Specific Server-wide Symmetry Encryption Key
```

## Configuration

The plugin is configured via Flask configuration options. The following table describes all available configuration options:

| ENV Variable  | Description  | Default Values |
|---|---|---|
| JOSE_CONTENT_TYPES | A list containing content types which might contain JWE requests.  | ['application/jose', 'application/jose+json'] |
| SERVER_PUB_JWK_ENDPOINT | A string defining the server endpoint that will provide the current elliptical ephemeral static public key (in JWT format). This endpoint is required in order to use the ECDH-ES algorithm. | '/serverpubkey' |
| JWE_ECDH_ES | Boolean defining ECDH-ES Support | True |
| ECDH_CURVE | A string defining the EC Curve Used for ECDH-ES | 'P-256' |
| JWE_ES_KEY_EXPIRES | ECDH-ES Public Key Expiration Time in Seconds.<br><br>For no expiration, set to -1 | -1 |
| JWE_SERVER_RSA_KEY | Shared RSA Key Used for Encryption (Must be of type jwkest.jwk.RSAKey) | None |
| JWE_SERVER_SYM_KEY | Shared Symmetric Key Used for Encryption (Must be of type jwkest.jwk.SYMKey) | None |
| JWE_REDIS_URI | Redis (or Redis Sentinel) URI String Used for Storing ECDH-ES Key (required for multiple, shared servers | 'redis://localhost:6379/1' |
| JWE_SERVER_KEY_PER_IP | Create ECDH-ES Key per IP | True |
| JWE_KEY_ENCRYPTION_KEY | 16, 24 or 32 byte Long to be used as AES-CBC key | None |
| JWE_SET_REQUEST_DATA | Replace request.data with decrypted request data | True

## Supported Algorithms and Encoding Methods

As of the time of this writing (module version 1.1.5), the supported algorithms are:

* ECDH-ES
* ECDH-ES+A128KW
* ECDH-ES+A192KW
* ECDH-ES+A256KW
* RSA1_5
* RSA-OAEP
* A128KW
* A192KW
* A256KW

**NOTE: RSA1_5 and RSA-OAEP used in ES mode is currently not supported.**

As of the time of this writing (module version 1.1.5), the supported encryption methods are:

* A128GCM
* A192GCM
* A256GCM
* A128CBC-HS256
* A192CBC-HS384
* A256CBC-HS512

## New Flask Request Functionality

### Function Members
| Function         | Description |
| ---------------- | ----------- |
| set_jwe_response(is_jwe) | Sets the response to use encrypted JWE (True) or send unencrypted content without JWE (False) |
| get_jwe_data() | Get Decrypted JWE Content |


### Data Members
| Data Member | Description |
| ----------- | ----------- |
| is_jwe      | Boolean field set to True if the request is a JWE request |
| jwe         | JWE Token Object | 

### Decorators
| Function             | Description |
| -------------------- | ----------- |
| jwe_request_required | Decorator to require a JWE request. Returns 400 error in case of non-JWE request | 

## Test Mode

When Flask's **app.config['TESTING']** config value is set to True, FlaskJWE adds a **POST** test endpoint. The hardcoded endpoint is currently '/flaskjwe-reverse-echo'. 
The endpoint will accept a JWE request and respond with a JWE response containing the reversed content.

CHANGELOG.md
# Flask-JWE Changelog

## 0.0.6 (2016-07-21)

Features:

  - Replace key XORing for security to plain AES key encryption in the Redis datastore
  - Add `alg` parameter for ES key retrieval to allow for EC and RSA keys

Refactoring:

  - Refactor Key Retrieval, Store and Building to simplify the FlaskJWE.get_server_key() function

Open Issues:

  - RSA-ES currently does not work as pyjwkest does not support saving CEK for JWE response processing (we are working to provide a pull request to fix this)
  - Due to this open issue, this version will not be published to PyPI

## 0.0.5 (2016-07-14)

Bugfixes:

  - Include appropriate `cty` JWE header in response JWE

## 0.0.4 (2016-07-14)

Bugfixes:

  - Remove `request._parsed_content_type` before calling request._parse_content_type() after Content-Type change

## 0.0.3 (2016-07-14)

Features:

  - Allow for JWE `cty` header to reset Content-Type for request after request data replacement

## 0.0.2 (2016-07-14)

Features:

  - Add support for Redis+Sentinel 
    
    `(redis+sentinel://server1:port,server2:port/serviceName)`
  - Add `@jwe_request_required` decorator to require a JWE request for an endpoint (otherwise, JWE requests are optional)
  - Add support for key per IP for ECDH-ES keys (turned on by default)    
  - Add support for internal ECDH-ES private key XOR value such that keys stored on Redis are not the actual keys used for ECDH-ES (turned off by default) 
  - Add support for Flask request data (*.data, .get_data(), .get_json()*) to be replaced with JWE's decrypted data (turned on by default)

PyPI:

  - Add `pyjwkest` dependency to setup.py

Dependencies:

  - Remove dependency `python-redis-lock`
  - Update `pyjwkest` version requirement to 1.1.7

## 0.0.1 (2016-02-23)

Features:

  - Allow any Flask endpoint to accept content using JWE requests and (*optionally, but by default*) return JWE responses
  - Provides support for all modes of JWE encryption as provided by `pyjwkest` 1.1.5  

Documentation & Examples:

  - Create initial README.md
  - Example of Flask-JWE use in test/functest_jwe.py

test/functest_jwe.py
__author__ = 'Matt David'

import logging
import os
import random
import requests
import unittest
import urllib

from Crypto.PublicKey import RSA

from flask import Flask, Response
from flask_jwe import FlaskJWE
from flask.ext.testing import LiveServerTestCase

from jwkest import jwe
from jwkest.jwe import JWE
from jwkest.jwk import KEYS, ECKey, RSAKey, SYMKey

log = logging.getLogger()

ECHO_TEST = "Echo Me In Reverse Please. I really just want to be echo'd so please do that"
ECHO_TEST_REVERSE = ECHO_TEST[::-1]

rsakey = RSA.generate(2048)
TEST_RSA_KEY = RSAKey(key=rsakey)
TEST_SYMKEY = SYMKey(KEY="My hollow echo", alg="HS512")

def index():
    return Response(status=200)

class FlaskJWEFunctionalTest(LiveServerTestCase):

    def create_app(self):

        random.seed()

        app = Flask(__name__)
        app.config['TESTING'] = True
        app.config['LIVESERVER_PORT'] = random.randrange(40000, 65000)
        app.config['DEBUG'] = True
        app.config['JWE_SERVER_RSA_KEY'] = TEST_RSA_KEY
        app.config['JWE_SERVER_SYM_KEY'] = TEST_SYMKEY
        app.config['JWE_KEY_ENCRYPTION_KEY'] = '68f94421ffa2fa97fc8230047e4b129a3e5ab6ad5bd88014e61473309aaae3e5'.decode('hex')
        app.config['JWE_ES_KEY_EXPIRES'] = 600
        app.config['JWE_REDIS_URI'] = 'redis://localhost:6379/4'
        flask_jwe = FlaskJWE(app)
        self.app = app

        # Add URL Route for /
        app.add_url_rule('/', 'index', index)
        return app

    def setUp(self):
        self.keys = [
            TEST_RSA_KEY,
            TEST_SYMKEY
        ]

    def test_direct_key_encryption(self):

        server_key = self.get_server_epk('ECDH-ES')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES',
            enc='A128GCM',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A128KW_keywrap_A128GCM_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A128KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A128KW',
            enc='A128GCM',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A192KW_keywrap_A128GCM_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A192KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A192KW',
            enc='A128GCM',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A256KW_keywrap_A128GCM_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A256KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A256KW',
            enc='A128GCM',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A128KW_keywrap_A192GCM_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A128KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A128KW',
            enc='A192GCM',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A128KW_keywrap_A256GCM_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A128KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A128KW',
            enc='A256GCM',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A128KW_keywrap_A128CBC_HS256_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A128KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A128KW',
            enc='A128CBC-HS256',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A128KW_keywrap_A192CBC_HS384_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A128KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A128KW',
            enc='A192CBC-HS384',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    def test_A128KW_keywrap_A256CBC_HS512_encryption(self):

        server_key = self.get_server_epk('ECDH-ES+A128KW')
        local_key = self.get_local_key(server_key)

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='ECDH-ES+A128KW',
            enc='A256CBC-HS512',
            epk=local_key,
            kid=server_key.kid,
            cty='text/plain'
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=[local_key])
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    # TODO: Followup with RSA-ES functionality once pyjwkest is updated to maintain CEK in JWE object
    def xtest_RSA_15_and_AESHMACSHA2_encryption(self):

        server_rsa_key = self.get_server_epk('RSA1_5')

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='RSA1_5',
            enc='A128CBC-HS256',
            cty='text/plain',
            kid=server_rsa_key.kid
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_rsa_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=self.keys)
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    # TODO: Followup with RSA-ES functionality once pyjwkest is updated to maintain CEK in JWE object
    def xtest_RSA_OAEP_and_A256GCM_encryption(self):

        server_rsa_key = self.get_server_epk('RSA-OAEP')

        submit_jwe = JWE(
            msg=ECHO_TEST,
            alg='RSA-OAEP',
            enc='A256GCM',
            cty='text/plain',
            kid=server_rsa_key.kid
        )

        # Encrypt and Send JWE
        serialized_jwe = submit_jwe.encrypt(keys=[server_rsa_key])
        jwe_response = requests.post('%s/flaskjwe-reverse-echo' % self.get_server_url(), serialized_jwe, headers={'Content-Type': 'application/jose'})

        # Decrypt JWE Response
        resp_jwe = jwe.factory(jwe_response.text)
        self.assertIsNotNone(resp_jwe)
        msg = resp_jwe.decrypt(keys=self.keys)
        self.assertEqual(ECHO_TEST_REVERSE, msg)

    # NOT YET SUPPORTED BY pyjwkest
    # - Direct Encryption "dir"
    # - PBES Encryption
    # - RSA-OAEP-256

    def get_server_epk(self, alg):

        pubkey_response = requests.get('%s/serverpubkey?%s' % (self.get_server_url(), urllib.urlencode({'alg': alg})),
                                       headers={'Accept': 'application/jose'})
        self.assertEqual(requests.codes.ok, pubkey_response.status_code)
        self.assertEqual('application/jose', pubkey_response.headers.get('content-type'))

        keys = KEYS()
        keys.load_dict(pubkey_response.json())
        if alg.startswith('ECDH-ES') and 'EC' in keys.key_types():
            return keys.as_dict()['EC'][0]
        elif alg.startswith('RSA') and 'RSA' in keys.key_types():
            return keys.as_dict()['RSA'][0]
        return None

    def get_local_key(self, epk):
        priv, pub = epk.curve.key_pair()
        return ECKey(crv=epk.curve.name(), x=pub[0], y=pub[1], d=priv, kid='%s-client' % epk.kid)

if __name__ == '__main__':
    unittest.main()


flask_jwe.py
import json
import random
import traceback

from base64 import b64encode, b64decode
from hashlib import sha256

from jwkest import jwe, base64_to_long
from jwkest.ecc import NISTEllipticCurve
from jwkest.jwe import JWE
from jwkest.jwk import ECKey, RSAKey
from flask import Response, request
from functools import wraps
from urlparse import urlparse

from Crypto import Random
from Crypto.Cipher import AES
from Crypto.PublicKey import RSA

try:
    from flask import _app_ctx_stack as stack
except ImportError:
    from flask import _request_ctx_stack as stack

try:
    from redis import StrictRedis, ConnectionError
    from redis.sentinel import Sentinel
except ImportError:
    pass

JOSE_CONTENT_TYPE = 'application/jose'
BS = AES.block_size
pad = lambda s: s + (BS - len(s) % BS) * chr(BS - len(s) % BS)
unpad = lambda s : s[:-ord(s[len(s)-1:])]

def jwe_request_required(f):
    @wraps(f)
    def wrapped(*args, **kwargs):
        if not request.__dict__.get('is_jwe', False):
            return Response('JWE Request Required', status=400, mimetype='text/plain')
        return f(*args, **kwargs)
    return wrapped

class FlaskJWE(object):
    def __init__(self, app=None):
        self.app = app
        if app is not None:
            self.init_app(app)

    def init_app(self, app):

        ##################
        # Set Defaults
        app.config.setdefault('JOSE_CONTENT_TYPES', ['application/jose', 'application/jose+json'])

        # Set Server Public Key Defaults
        app.config.setdefault('SERVER_PUB_JWK_ENDPOINT', '/serverpubkey')
        app.config.setdefault('JWE_ECDH_ES', True)
        app.config.setdefault('ECDH_CURVE', 'P-256')  # Available Options: ['P-256', 'P-384', 'P-521']
        app.config.setdefault('JWE_ES_KEY_EXPIRES', -1) # Expiration Time in Seconds
        app.config.setdefault('JWE_SERVER_RSA_KEY', None)
        app.config.setdefault('JWE_SERVER_SYM_KEY', None)
        app.config.setdefault('JWE_SERVER_KEY_PER_IP', True)
        app.config.setdefault('JWE_KEY_ENCRYPTION_KEY', None)
        app.config.setdefault('JWE_SET_REQUEST_DATA', True)

        # Set Redis Defaults
        app.config.setdefault('JWE_REDIS_URI', 'redis://localhost:6379/1')

        # Add Public Key Endpoint
        if 'JWE_ECDH_ES' in app.config and app.config['JWE_ECDH_ES']:
            if not self.route_in_use(app, app.config['SERVER_PUB_JWK_ENDPOINT']):
                app.add_url_rule(app.config['SERVER_PUB_JWK_ENDPOINT'], 'jwkendpoint', self.return_jwk_pub)
            else:
                app.logger.error("Unable to Create Server EC Public Key JWK Endpoint")

        # If TESTING mode is True, setup reversing echo endpoint
        if 'TESTING' in app.config and app.config['TESTING']:
            app.add_url_rule('/flaskjwe-reverse-echo', 'test-reverse-echo', self.reverse_echo, methods=['POST'])

        # Setup Before / After Request Processing for JOSE/JWE Request
        app.before_request(self.on_request_start)
        app.after_request(self.on_request_end)

    def get_keys(self, alg):
        keys = []
        if self.app.config['JWE_SERVER_RSA_KEY']:
            keys.append(self.app.config['JWE_SERVER_RSA_KEY'])
        if self.app.config['JWE_SERVER_SYM_KEY']:
            keys.append(self.app.config['JWE_SERVER_SYM_KEY'])

        server_es_key = self.get_server_key(self.app, alg)
        if server_es_key:
            keys.append(server_es_key)

        return keys

    def on_request_start(self):

        def set_jwe_response(is_jwe):
            request.is_jwe = is_jwe

        request.jwe = self.is_jwe(request)
        if request.jwe:
            request.set_jwe_response = set_jwe_response
            self.jwe_decrypt(request.jwe)

    def on_request_end(self, response_class):

        if not request.__dict__.get('is_jwe', False):
            return response_class

        # Re-encrypt JWE
        return_jwe = self.jwe_encrypt(response_class)
        response_class.set_data(return_jwe)
        response_class.mimetype = 'application/jose'
        response_class.content_type = 'application/jose'
        return response_class

    def is_jwe(self, req):

        if req.content_type.lower() not in self.app.config.get('JOSE_CONTENT_TYPES'):
            return False

        return jwe.factory(req.get_data())

    def jwe_encrypt(self, response_class):

        jwe = JWE(
            msg=response_class.get_data(),
            alg=request.jwe.jwt.headers.get('alg'),
            enc=request.jwe.jwt.headers.get('enc'),
            cty=response_class.content_type
        )

        # Add EPK and KID for ECDH-ES Requests
        keys = []
        if jwe.alg.startswith('ECDH-ES'):
            jwe._dict['kid'] = request.jwe.jwt.headers.get('epk', {}).get('kid')
            jwe._dict['epk'] = self.get_server_key(self.app, request.jwe.jwt.headers.get('alg'))
            keys.append(ECKey(**request.jwe.jwt.headers.get('epk')))

        elif jwe.alg.startswith('RSA'):
            # TODO: Finish RSA-ES Once Two-way JWS Questions are Resolved
            pass

        return jwe.encrypt(keys=keys)

    def jwe_decrypt(self, jwe):
        try:
            # Decrypt JWE
            msg = jwe.decrypt(keys=self.get_keys(jwe.jwt.headers.get('alg')))

            def get_jwe_data():
                return msg

            request.is_jwe = True
            request.get_jwe_data = get_jwe_data
            if self.app.config['JWE_SET_REQUEST_DATA']:
                request.data = msg
                request._cached_data = msg
                if jwe.jwt.headers.get('cty'):
                    request.environ['CONTENT_TYPE'] = jwe.jwt.headers.get('cty')
                    if '_parsed_content_type' in request.__dict__:
                        del request.__dict__['_parsed_content_type']
                    request._parse_content_type()
        except Exception as e:
            self.app.logger.error("Unable to decrypt JWE: %s" % str(e))
            self.app.logger.error(traceback.format_exc())
            return Response(json.dumps({'error_message': 'Unable to decrypt JWE Token'}), status=500, mimetype='application/json')

    def return_jwk_pub(self):

        alg = request.args.get('alg', None)
        if not alg:
            return Response('Required parameter "alg" missing', status=400, mimetype='text/plain')

        ret = {'keys': []}

        fullkey = self.get_server_key(self.app, alg)
        if fullkey:
            ret['keys'].append(fullkey.serialize(False))
        return Response(json.dumps(ret), content_type=JOSE_CONTENT_TYPE)

    def route_in_use(self, app, endpoint):
        for rule in app.url_map.iter_rules():
            if endpoint == rule.endpoint:
                return True
        return False

    #########################################
    # Server Ephemeral Static Key Operations
    #########################################
    def get_server_key(self, app, alg):

        try:
            key_name = 'flask-jwe-%s-key' % alg.lower()
            if app.config['JWE_SERVER_KEY_PER_IP']:
                key_name += '-%s' % self.get_remote_ip()

            key = self.get_redis_jwk(key_name)
            if not key:
                app.logger.debug('Creating New %s KeyPair' % alg)
                key = self.set_redis_jwk(key_name, self.generate_jwk(sha256(key_name).hexdigest(), alg))

        except Exception as e:
            app.logger.error('Exception Occurred Generating Retrieving / Setting ES KeyPair: %s' % str(e))
            return None

        return key

    @jwe_request_required
    def reverse_echo(self):
        if self.app.config['JWE_SET_REQUEST_DATA']:
            ret_str = request.get_data()[::-1]
        else:
            ret_str = request.get_jwe_data()[::-1]
        return Response(ret_str)

    #########################################
    # Utility Functionality
    #########################################
    def generate_jwk(self, kid, alg):

        ret_jwk = None

        if alg.startswith('ECDH-ES'):
            curve = NISTEllipticCurve.by_name(self.app.config['ECDH_CURVE'])
            priv, pub = curve.key_pair()
            ret_jwk = ECKey(use='enc', x=pub[0], y=pub[1], d=priv, crv=self.app.config['ECDH_CURVE'], kid=kid)

        elif alg.startswith('RSA'):
            new_rsa_key = RSA.generate(2048)
            ret_jwk = RSAKey(use='enc', key=new_rsa_key, kid=kid)

        return ret_jwk

    def build_jwk(self, serialized_jwk):

        jwk_dict = json.loads(serialized_jwk)

        if jwk_dict.get('kty') == 'EC':
            return ECKey(**jwk_dict)
        elif jwk_dict.get('kty') == 'RSA':
            return RSAKey(**jwk_dict)
        return None

    def get_redis_jwk(self, key_name):

        client = self.get_redis_client(connection_uri=self.app.config['JWE_REDIS_URI'])
        if not client:
            self.app.logger.fatal('Unable to connect to Redis')
            return None

        value = client.get(key_name)
        if not value:
            return None

        if self.app.config['JWE_ES_KEY_EXPIRES'] > 0 and client.ttl(key_name) < 30:
            client.expire(key_name, self.app.config['JWE_ES_KEY_EXPIRES'])

        if not self.app.config['JWE_KEY_ENCRYPTION_KEY']:
            return self.build_jwk(value)

        enc = b64decode(value)
        cipher = AES.new(self.app.config['JWE_KEY_ENCRYPTION_KEY'], AES.MODE_CBC, enc[:16])
        value = unpad(cipher.decrypt(enc[16:]))

        return self.build_jwk(value)

    def set_redis_jwk(self, key_name, jwk):

        client = self.get_redis_client(connection_uri=self.app.config['JWE_REDIS_URI'])
        if not client:
            self.app.logger.fatal('Unable to connect to Redis')
            return None

        if self.app.config['JWE_KEY_ENCRYPTION_KEY']:

            if len(self.app.config['JWE_KEY_ENCRYPTION_KEY']) not in [16, 24, 32]:
                self.app.logger.fatal('JWE_KEY_ENCRYPTION_KEY Must be 16, 24 or 32 bytes long')
                return None

            raw = pad(json.dumps(jwk.serialize(True)))
            iv = Random.new().read(AES.block_size)
            cipher = AES.new(self.app.config['JWE_KEY_ENCRYPTION_KEY'], AES.MODE_CBC, iv)
            key_value = b64encode(iv + cipher.encrypt(raw))
        else:
            key_value = json.dumps(jwk.serialize(True))

        client.setnx(key_name, key_value)

        return self.get_redis_jwk(key_name)

    #########################################
    # Utility Functionality
    #########################################
    def get_remote_ip(self):
        if request.access_route:
            return request.access_route[0]
        else:
            return request.remote_addr or '127.0.0.1'

    def get_redis_client(self, connection_uri=None, read_only=False):

        if not connection_uri:
            self.app.logger.fatal('Redis Connection URI Required')
            return None

        redis_client = None
        redis_uri = urlparse(connection_uri)

        if not StrictRedis or not Sentinel:
            self.app.logger.fatal('Unable to Import Redis')

        if redis_uri.scheme == 'redis':

            redis_args = {
                'host': redis_uri.hostname,
                'port': redis_uri.port
            }
            if redis_uri.path.lstrip('/'):
                redis_args['db'] = redis_uri.path.lstrip('/')

            redis_client = StrictRedis(**redis_args)

        elif redis_uri.scheme == 'redis+sentinel':

            sentinel_endpoints = [(y[0], y[1]) for y in [x.strip().split(':') for x in redis_uri.netloc.split(',')]]
            service_name = redis_uri.path.lstrip('/')
            sentinel = Sentinel(sentinel_endpoints, socket_timeout=2, retry_on_timeout=True)

            for _ in range(5):
                if read_only:
                    redis_client = sentinel.slave_for(service_name)
                else:
                    redis_client = sentinel.master_for(service_name)

                try:
                    redis_client.info()
                    break
                except ConnectionError as e:
                    self.app.logger.warn('Redis %s Sentinel Connection Error: %s' % ('Slave' if read_only else 'Master', str(e)))
                    redis_client = None

        else:
            self.app.logger.fatal('Unsupported REDIS URI Scheme. Unable to get Redis client [SCHEME: %s]' % redis_uri.scheme)

        return redis_client